# О MovieLens
MovieLens — это веб-система рекомендаций и виртуальное сообщество, которое рекомендует пользователям фильмы для просмотра на основе их предпочтений с помощью совместной фильтрации оценок и рецензий членов сообщества. Она содержит около 11 миллионов оценок примерно 8500 фильмов. MovieLens был создан в 1997 году GroupLens Research, исследовательской лабораторией факультета компьютерных наук и инженерии Университета Миннесоты, с целью сбора исследовательских данных и персонализированных рекомендаций.

## История
MovieLens не была первой системой рекомендаций, созданной GroupLens. В мае 1996 года GroupLens создала коммерческое предприятие под названием Net Perceptions, которое обслуживало таких клиентов, как E! Online и Amazon.com. E! Online использовала услуги Net Perceptions для создания системы рекомендаций для Moviefinder.com, а Amazon.com использовала технологию компании для формирования своей ранней системы рекомендаций для покупок потребителей.
Когда в 1997 году закрылся другой сайт с рекомендациями фильмов, eachmovie.org, исследователи, которые его создали, публично обнародовали 
анонимные данные собранных рейтингов, чтобы другие исследователи могли их использовать. Исследовательская группа GroupLens под руководством Брента Далена и Джона Херлокера использовала этот набор данных для запуска нового сайта рекомендаций фильмов, который они назвали MovieLens.

# О тестовом наборе данных

Мы будем работать с сокращенным набором данных MovieLens (ml-latest-small). Он содержит 5-звездочные рейтинги и теги, созданные 610 пользователями в период с 1996 по 2018 год: всего 100836 оценок и 3683 тегов для 9742 фильмов. Данные содержатся в четырёх файлах: links.csv, movies.csv, ratings.csv и tags.csv. Для их анализа будем использовать методы внутри классов Links, Movies, Ratings и Tags, соответственно.

Проведем анализ первой тысячи записей для каждого файла.

# Отчет об анализе MovieLens

Все необходимые инструменты для анализа данных MovieLens содержатся в модуле movielens_analysis.  
Нам понадобятся классы Movies, Ratings, Tags, Links. Импортируем их:

In [1]:
%timeit -n 1 -r 1 from movielens_analysis import Movies, Ratings, Tags, Links, attr_list
from movielens_analysis import Movies, Ratings, Tags, Links, attr_list

146 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


## Часть 1. Класс Movies

Создадим экземпляр класса:

In [2]:
%timeit -n 1 -r 1 movies = Movies('../datasets/movies.csv', 1000)
movies = Movies('../datasets/movies.csv', 1000)

5.77 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Атрибуты класса:

In [3]:
%timeit -n 1 -r 1 attr_list(movies)
attr_list(movies)

5.59 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['N', 'COLS', 'movieId', 'title', 'genres', 'year', 'movie_index']

N - количество фильмов, COLS - названия колонок данных. movieId, title, genres, year - параллельные списки movieId, названий, жанров и годов выхода фильмов. Инициализатор запускает функцию-генератор parse_csv в цикле, затем преобразует строковые значения в соответствующий тип (movieId: int, genres - list[str] и т.д.). movie_index - словарь {movieId: index} для поиска данных из списков title, genres, year и т.д. по movieId.

Функция **parse_csv** - генератор. Считывает построчно файл csv и разбивает строку с помощью функции qsplit на колонки. Выводит список значений в виде строк.

Функция **qsplit** разбивает строку через запятую, с учетом кавычек. Проходит строку посимвольно и запоминает, внутри кавычек мы или снаружи. Запятые внутри кавычек игнорируются, кавычки удаляются с помощью простой функции qshave (тернарный оператор if (есть кавычки) обрезать else оставить).

#### 1. Когда выходило больше всего фильмов?

Метод **dist_by_release** использует collections.Counter для подсчета элементов списка self.year и выводит в порядке убывания количества фильмов для каждого года с помощью метода most_common.

In [4]:
%timeit -n 1 -r 1 movies.dist_by_release()
movies.dist_by_release()

219 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1995: 224,
 1994: 184,
 1996: 181,
 1993: 101,
 1992: 23,
 1990: 15,
 1991: 15,
 1989: 14,
 1986: 9,
 1982: 8,
 1940: 8,
 1957: 8,
 1987: 8,
 1980: 8,
 1981: 7,
 1988: 7,
 1979: 7,
 1955: 6,
 1959: 6,
 1968: 6,
 1997: 6,
 1939: 6,
 1985: 6,
 1967: 5,
 1965: 5,
 1951: 5,
 1958: 5,
 1944: 5,
 1941: 5,
 1975: 5,
 1971: 5,
 1984: 5,
 1964: 4,
 1973: 4,
 1954: 4,
 1934: 4,
 1960: 4,
 1963: 4,
 1950: 4,
 1974: 4,
 1983: 4,
 1977: 3,
 1937: 3,
 1972: 3,
 1952: 3,
 1961: 3,
 1953: 3,
 1946: 3,
 1938: 3,
 1956: 3,
 1962: 3,
 1976: 2,
 1969: 2,
 1970: 2,
 1942: 2,
 1945: 2,
 1947: 2,
 1935: 2,
 1936: 2,
 1949: 2,
 1978: 2,
 1943: 1,
 1932: 1,
 1966: 1,
 1948: 1,
 1933: 1,
 1931: 1}

Больше всего фильмов снято в 1995 году, но после 1996 года вообще нет данных. Следовательно, по первым 1000 фильмов статистика неполная. Повторим анализ по всему датасету:

In [5]:
%timeit -n 1 -r 1 movies_full = Movies('../datasets/movies.csv', 10000)
movies_full = Movies('../datasets/movies.csv', 10000)

61 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Сколько же в нём фильмов?

In [6]:
len(movies_full.movieId)

9742

In [7]:
%timeit -n 1 -r 1 movies_full.dist_by_release()
movies_full.dist_by_release()

574 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{2002: 311,
 2006: 295,
 2001: 294,
 2007: 284,
 2000: 283,
 2009: 282,
 2003: 279,
 2004: 279,
 2014: 278,
 1996: 276,
 2015: 274,
 2005: 273,
 2008: 269,
 1999: 263,
 1997: 260,
 1995: 259,
 1998: 258,
 2011: 254,
 2010: 247,
 2013: 239,
 1994: 237,
 2012: 233,
 2016: 218,
 1993: 198,
 1992: 167,
 1988: 165,
 1987: 153,
 1990: 147,
 1991: 147,
 2017: 147,
 1989: 142,
 1986: 139,
 1985: 126,
 1984: 101,
 1981: 92,
 1980: 89,
 1982: 87,
 1983: 83,
 1979: 69,
 1977: 63,
 1973: 59,
 1978: 59,
 1965: 47,
 1971: 47,
 1974: 45,
 1976: 44,
 1964: 43,
 1967: 42,
 1968: 42,
 1975: 42,
 1966: 42,
 2018: 41,
 1962: 40,
 1972: 39,
 1963: 39,
 1959: 37,
 1960: 37,
 1955: 36,
 1969: 35,
 1961: 34,
 1970: 33,
 1957: 33,
 1958: 31,
 1953: 30,
 1956: 30,
 1940: 25,
 1949: 25,
 1954: 23,
 1942: 23,
 1939: 23,
 1946: 23,
 1951: 22,
 1950: 21,
 1947: 20,
 1948: 20,
 1941: 18,
 1936: 18,
 1945: 17,
 1937: 16,
 1952: 16,
 1944: 16,
 1938: 15,
 1931: 14,
 1935: 13,
 None: 13,
 1933: 12,
 1934: 11,
 1943: 10

Из более полного датасета видно, что больше всего фильмов снималось в первое десятилетие XXI века. Интересно, что это за фильмы вышли в 1902 и 1903 годах?

#### 2. В каких жанрах снято больше всего фильмов?

Метод **dist_by_genres** использует collections.Counter для подсчета элементов списка списков self.genres, и выводит в порядке убывания количества фильмов для каждого жанра с помощью метода most_common.

In [8]:
%timeit -n 1 -r 1 movies.dist_by_genres()
movies.dist_by_genres()

397 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Drama': 507,
 'Comedy': 365,
 'Romance': 208,
 'Thriller': 179,
 'Action': 158,
 'Adventure': 126,
 'Crime': 122,
 'Children': 100,
 'Fantasy': 69,
 'Sci-Fi': 69,
 'Mystery': 58,
 'Musical': 53,
 'Horror': 51,
 'War': 48,
 'Animation': 37,
 'Documentary': 25,
 'Western': 23,
 'Film-Noir': 18,
 'IMAX': 3}

Львиную долю составляют драма и комедия, на третьем месте - романтическое кино. Триллеры и боевики тоже довольно популярны, а вот научной фантастики мало, что неудивительно: в кино этот жанр набрал популярность в последние 20-30 лет благодаря спецэффектам. Поэтому вернемся к нашему полному датасету, который включает фильмы 2000-х годов:

In [9]:
%timeit -n 1 -r 1 movies_full.dist_by_genres()
movies_full.dist_by_genres()

4.81 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Drama': 4361,
 'Comedy': 3756,
 'Thriller': 1894,
 'Action': 1828,
 'Romance': 1596,
 'Adventure': 1263,
 'Crime': 1199,
 'Sci-Fi': 980,
 'Horror': 978,
 'Fantasy': 779,
 'Children': 664,
 'Animation': 611,
 'Mystery': 573,
 'Documentary': 440,
 'War': 382,
 'Musical': 334,
 'Western': 167,
 'IMAX': 158,
 'Film-Noir': 87}

В более полном датасете Sci-Fi переместился с 10 на 8 место, и появился жанр Fantasy. А вот романтическое кино сместилось с 3 на 5 место. Комедии, триллеры и боевики по-прежнему популярны. Популярность драматического кино объясняется тем, что это очень обширное понятие и включает в себя мелодрамы, исторические драмы и множество других поджанров.

#### 3. Сколько может быть жанров у одного фильма?

Метод **most_genres** объединяет список self.title и генератор количества жанров map(len, self.genres) с помощью zip, сортирует по убыванию количества жанров и преобразует в словарь.

In [10]:
%timeit -n 1 -r 1 movies.most_genres(5)
movies.most_genres(5)

581 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Aladdin and the King of Thieves (1996)': 6,
 'Space Jam (1996)': 6,
 'All Dogs Go to Heaven 2 (1996)': 6,
 'Beauty and the Beast (1991)': 6,
 'Super Mario Bros. (1993)': 6}

И, по традиции, для полного датасета:

In [11]:
%timeit -n 1 -r 1 movies_full.most_genres(5)
movies_full.most_genres(5)

6.43 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Rubber (2010)': 10,
 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8,
 'Mars Needs Moms (2011)': 7,
 'Tangled (2010)': 7,
 'Inception (2010)': 7}

Нетрудно заметить, что многие фильмы отнесены сразу к нескольким жанрам.

#### 4. Поиск по словам в названии фильма

Метод **title_search** принимает список слов и возвращает словарь {movieId: название_фильма}. Метод проходит циклом по списку названий фильмов self.title и проверяет, что все слова встречаются в названии.

In [12]:
%timeit -n 1 -r 1 movies.title_search(['star', 'wars'])
movies.title_search(['star', 'wars'])

1.08 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{260: 'Star Wars: Episode IV - A New Hope (1977)',
 1196: 'Star Wars: Episode V - The Empire Strikes Back (1980)',
 1210: 'Star Wars: Episode VI - Return of the Jedi (1983)'}

Какой-то скудный список получился, повторим с полным датасетом:

In [13]:
%timeit -n 1 -r 1 movies_full.title_search(['star', 'wars'])
movies_full.title_search(['star', 'wars'])

7.31 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{260: 'Star Wars: Episode IV - A New Hope (1977)',
 1196: 'Star Wars: Episode V - The Empire Strikes Back (1980)',
 1210: 'Star Wars: Episode VI - Return of the Jedi (1983)',
 2628: 'Star Wars: Episode I - The Phantom Menace (1999)',
 5378: 'Star Wars: Episode II - Attack of the Clones (2002)',
 33493: 'Star Wars: Episode III - Revenge of the Sith (2005)',
 61160: 'Star Wars: The Clone Wars (2008)',
 79006: "Empire of Dreams: The Story of the 'Star Wars' Trilogy (2004)",
 122886: 'Star Wars: Episode VII - The Force Awakens (2015)',
 135216: 'The Star Wars Holiday Special (1978)',
 166528: 'Rogue One: A Star Wars Story (2016)',
 179819: 'Star Wars: The Last Jedi (2017)',
 187595: 'Solo: A Star Wars Story (2018)'}

Поскольку год указан в названии фильма, можно этим же методом искать по годам. И мы наконец можем узнать, что за фильмы снимались в 1902 и 1903 годах.

In [14]:
%timeit -n 1 -r 1 movies_full.title_search(['(1902)'])
movies_full.title_search(['(1902)'])

6.45 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{32898: 'Trip to the Moon, A (Voyage dans la lune, Le) (1902)'}

«Путешествие на Луну» (фр. Le Voyage dans la Lune) — французский научно-фантастический фильм 1902 года, режиссёра Жоржа Мельеса. Немая короткометражная фарсовая комедия, пародирующая сюжеты романов Жюля Верна «Из пушки на Луну» и Герберта Уэллса «Первые люди на Луне». Поставлена Мельесом по собственному сценарию, на созданных им декорациях, на его частной студии и силами его актёрской группы. Считается первым научно-фантастическим фильмом в истории кино.

In [15]:
%timeit -n 1 -r 1 movies_full.title_search(['(1903)'])
movies_full.title_search(['(1903)'])

6.99 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{49389: 'The Great Train Robbery (1903)'}

«Большое ограбление поезда» (англ. The Great Train Robbery) — американский художественный немой фильм Эдвина Портера, снятый на студии Томаса Эдисона в 1903 году. Один из самых первых холд-ап вестернов, положивший начало целому жанру в американском кино.

## Часть 2. Класс Ratings

Создадим экземпляр класса:

In [16]:
%timeit -n 1 -r 1 ratings = Ratings('../datasets/ratings.csv', 1000)
ratings = Ratings('../datasets/ratings.csv', 1000)

51.5 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Атрибуты класса:

In [17]:
%timeit -n 1 -r 1 attr_list(ratings)
attr_list(ratings)

6.43 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['N', 'COLS', 'userId', 'movieId', 'rating', 'year', 'movies', 'users']

N - количество фильмов, COLS - названия колонок данных. userId, movieId, rating, year - параллельные списки userId, movieId, оценок и годов выставления оценок.  
movies и users - автоматически созданные экземпляры вложенных классов Movies и Users. При создании им передается ссылка на экземпляр внешнего класса Movies, созданного из файла movies.csv. Файл movies.csv должен находиться в одной папке с ratings.csv  
Метод считывает и преобразует данные из ratings.csv аналогично методу Movies (см. Часть 1).

Атрибуты классов Ratings.Movies и Ratings.Users(Ratings.Movies):

In [18]:
%timeit -n 1 -r 1 attr_list(ratings.movies)
attr_list(ratings.movies)

6.21 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['userId', 'movieId', 'rating', 'year', 'movies', 'movie_ratings']

In [19]:
%timeit -n 1 -r 1 attr_list(ratings.users)
attr_list(ratings.users)

4.64 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['userId', 'movieId', 'rating', 'year', 'movies', 'movie_ratings']

Кроме уже известных нам userId, movieId, rating и year, атрибут movies содержит автоматически созданный экземпляр класса Movies с информациях о названиях фильмов. Файл movies.csv должен находиться в одной папке с файлом ratings.csv. Атрибут movie_ratings - словарь {movieId: список_оценок} для быстрого поиска рейтинга фильма; создается с помощью setdefault(movieId, []).append(rating) для каждой пары movieId: rating.

## Часть 2.1. Класс Ratings.Movies

Содержит статистику отзывов по фильмам.

#### 1. Когда пользователи чаще всего оценивали фильмы?
Метод **dist_by_year** использует collections.Counter для подсчета элементов списка self.year, сортирует и выводит в порядке возрастания года выставления оценки.

In [20]:
%timeit -n 1 -r 1 ratings.movies.dist_by_year()
ratings.movies.dist_by_year()

80 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1996: 358,
 1999: 82,
 2000: 296,
 2001: 70,
 2005: 121,
 2006: 4,
 2007: 1,
 2011: 39,
 2015: 29}

Выборка снова выглядит странно, ведь в 1996 году было гораздо меньше пользователей интернета, чем в 2010-х. Повторим с более полной версией датасета:

In [21]:
%timeit -n 1 -r 1 ratings_full = Ratings('../datasets/ratings.csv', 1_000_000)
ratings_full = Ratings('../datasets/ratings.csv', 1_000_000)

296 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Сколько оценок мы сейчас загрузили?

In [22]:
len(ratings_full.userId)

100836

In [23]:
%timeit -n 1 -r 1 ratings_full.movies.dist_by_year()
ratings_full.movies.dist_by_year()

7.38 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1996: 6040,
 1997: 1916,
 1998: 507,
 1999: 2439,
 2000: 10061,
 2001: 3922,
 2002: 3478,
 2003: 4014,
 2004: 3279,
 2005: 5813,
 2006: 4059,
 2007: 7114,
 2008: 4351,
 2009: 4158,
 2010: 2300,
 2011: 1690,
 2012: 4657,
 2013: 1664,
 2014: 1439,
 2015: 6616,
 2016: 6702,
 2017: 8199,
 2018: 6418}

Распределение по годам более равномерное, но какие-либо выводы сделать сложно: сказывается неполная выборка. Полная версия MovieLens насчитывает около 33 миллионов оценок.

#### 2. Какие оценки чаще ставят пользователи?
Метод **dist_by_rating** использует collections.Counter для подсчета элементов списка self.rating, сортирует и выводит в порядке возрастания оценки.

In [24]:
%timeit -n 1 -r 1 ratings.movies.dist_by_rating()
ratings.movies.dist_by_rating()

105 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{0.5: 24,
 1.0: 39,
 1.5: 11,
 2.0: 57,
 2.5: 7,
 3.0: 253,
 3.5: 17,
 4.0: 292,
 4.5: 33,
 5.0: 267}

Пользователи более склонны оставлять хорошие оценки, чаще целочисленные, а не дробные.

#### 3. Топ фильмов по количеству оценок.
Метод **top_by_num_of_ratings** вызывает метод self.top_by_ratings с параметром metric='count'.

In [25]:
%timeit -n 1 -r 1 ratings.movies.top_by_num_of_ratings(10)
ratings.movies.top_by_num_of_ratings(10)

1.7 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Usual Suspects, The (1995)': 4,
 'Pulp Fiction (1994)': 4,
 'Fugitive, The (1993)': 4,
 "Schindler's List (1993)": 4,
 'Batman (1989)': 4,
 'Silence of the Lambs, The (1991)': 4,
 'Fargo (1996)': 4,
 'Aladdin (1992)': 4,
 'Beauty and the Beast (1991)': 4,
 'Toy Story (1995)': 3}

Маловато оценок для самых популярных фильмов. И где Звездные войны?

In [26]:
%timeit -n 1 -r 1 ratings_full.movies.top_by_num_of_ratings(30)
ratings_full.movies.top_by_num_of_ratings(30)

68.7 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Forrest Gump (1994)': 329,
 'Shawshank Redemption, The (1994)': 317,
 'Pulp Fiction (1994)': 307,
 'Silence of the Lambs, The (1991)': 279,
 'Matrix, The (1999)': 278,
 'Star Wars: Episode IV - A New Hope (1977)': 251,
 'Jurassic Park (1993)': 238,
 'Braveheart (1995)': 237,
 'Terminator 2: Judgment Day (1991)': 224,
 "Schindler's List (1993)": 220,
 'Fight Club (1999)': 218,
 'Toy Story (1995)': 215,
 'Star Wars: Episode V - The Empire Strikes Back (1980)': 211,
 'Usual Suspects, The (1995)': 204,
 'American Beauty (1999)': 204,
 'Seven (a.k.a. Se7en) (1995)': 203,
 'Independence Day (a.k.a. ID4) (1996)': 202,
 'Apollo 13 (1995)': 201,
 'Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)': 200,
 'Lord of the Rings: The Fellowship of the Ring, The (2001)': 198,
 'Star Wars: Episode VI - Return of the Jedi (1983)': 196,
 'Godfather, The (1972)': 192,
 'Fugitive, The (1993)': 190,
 'Batman (1989)': 189,
 'Saving Private Ryan (1998)': 188,
 'Lord of the Rings

Перечислены действительно очень известные фильмы. Но при этом совсем нет новых фильмов, хотя в датасете они есть. Это можно было бы понять из возрастного профиля пользователей, но данная информация недоступна.

#### 4. Топ фильмов по среднему рейтингу
Метод **top_by_ratings** использует метод self._top_x_by_y для применения метрики к списку оценок self.rating и сортировки списка названий по величине метрики. Список self.movieId преобразуется в список названий фильмов с помощью метода self._title_by_Id.

Метод **_top_x_by_y** - абстракция для вычисления топ-n значений x (например, названий фильмов) по заданной метрике от списка свойств y (например, оценок). Создает collections.defaultdict с элементом списка x в качестве ключа, и списком значений из списка y для данного x.  
Затем применяет данную метрику к списку значений, сортирует и выводит словарь {x: value} по убыванию значения метрики value для данного x.  
Для вычисления метрик используются методы self._average, self._median, self._pvariance (population variance), вычисляющие метрики по известным формулам.

In [27]:
%timeit -n 1 -r 1 ratings.movies.top_by_ratings(10)
ratings.movies.top_by_ratings(10)

1.47 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Star Wars: Episode IV - A New Hope (1977)': 5.0,
 'Wizard of Oz, The (1939)': 5.0,
 'Ghost and the Darkness, The (1996)': 5.0,
 "Monty Python's Life of Brian (1979)": 5.0,
 'Monty Python and the Holy Grail (1975)': 5.0,
 'Princess Bride, The (1987)': 5.0,
 'Terminator, The (1984)': 5.0,
 'Back to the Future (1985)': 5.0,
 'Fantasia (1940)': 5.0,
 'Jungle Book, The (1967)': 5.0}

И для более полной выборки:

In [28]:
%timeit -n 1 -r 1 ratings_full.movies.top_by_ratings(10)
ratings_full.movies.top_by_ratings(10)

70.9 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Lesson Faust (1994)': 5.0,
 'Lamerica (1994)': 5.0,
 "Jonah Who Will Be 25 in the Year 2000 (Jonas qui aura 25 ans en l'an 2000) (1976)": 5.0,
 'Heidi Fleiss: Hollywood Madam (1995)': 5.0,
 'Come and See (Idi i smotri) (1985)': 5.0,
 'Enter the Void (2009)': 5.0,
 'Belle époque (1992)': 5.0,
 'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0}

Все перечисленные фильмы имеют среднюю оценку 5.00. Скорее всего, целочисленное среднее свидетельствует о малом количестве оценок.

#### 5. Топ фильмов по медианному рейтингу

In [29]:
%timeit -n 1 -r 1 ratings.movies.top_by_ratings(10, metric='median')
ratings.movies.top_by_ratings(10, metric='median')

1.65 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Fugitive, The (1993)': 5.0,
 'Star Wars: Episode IV - A New Hope (1977)': 5.0,
 'Tommy Boy (1995)': 5.0,
 'Forrest Gump (1994)': 5.0,
 'Jurassic Park (1993)': 5.0,
 'Tombstone (1993)': 5.0,
 'Dances with Wolves (1990)': 5.0,
 'Pinocchio (1940)': 5.0,
 'Star Wars: Episode V - The Empire Strikes Back (1980)': 5.0,
 'American Beauty (1999)': 5.0}

Увеличим выборку:

In [30]:
%timeit -n 1 -r 1 ratings_full.movies.top_by_ratings(10, metric='median')
ratings_full.movies.top_by_ratings(10, metric='median')

81.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{"Guess Who's Coming to Dinner (1967)": 5.0,
 'Secrets & Lies (1996)': 5.0,
 'Band of Brothers (2001)': 5.0,
 'Angels and Insects (1995)': 5.0,
 "Man Bites Dog (C'est arrivé près de chez vous) (1992)": 5.0,
 "Swept Away (Travolti da un insolito destino nell'azzurro mare d'Agosto) (1975)": 5.0,
 'Trial, The (Procès, Le) (1962)': 5.0,
 'Last Tango in Paris (Ultimo tango a Parigi) (1972)': 5.0,
 'Memories of Murder (Salinui chueok) (2003)': 5.0,
 'Yi Yi (2000)': 5.0}

Все перечисленные фильмы имеют медианную оценку 5.00, но количество оценок может быть любым.

#### 6. Топ фильмов с неоднозначным рейтингом
Метод **top_controversial** вызывает метод self.top_by_ratings с параметром metric='variance'.

In [31]:
%timeit -n 1 -r 1 ratings.movies.top_controversial(10)
ratings.movies.top_controversial(10)

2.33 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Bambi (1942)': 5.06,
 'Rescuers, The (1977)': 5.06,
 'My Fair Lady (1964)': 5.06,
 'Matrix, The (1999)': 4.0,
 "Schindler's List (1993)": 3.42,
 'Good Will Hunting (1997)': 3.06,
 'Courage Under Fire (1996)': 3.06,
 'Usual Suspects, The (1995)': 2.42,
 'Dazed and Confused (1993)': 2.25,
 'Fight Club (1999)': 2.25}

И для более полной выборки оценок:

In [32]:
%timeit -n 1 -r 1 ratings_full.movies.top_controversial(30)
ratings_full.movies.top_controversial(30)

91.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{"Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)": 5.06,
 'Fanny and Alexander (Fanny och Alexander) (1982)': 5.06,
 'Troll 2 (1990)': 4.5,
 'Lassie (1994)': 4.0,
 'Zed & Two Noughts, A (1985)': 4.0,
 'Kwaidan (Kaidan) (1964)': 4.0,
 'Emma (2009)': 4.0,
 'Play Time (a.k.a. Playtime) (1967)': 3.72,
 'Room, The (2003)': 3.56,
 'Peeping Tom (1960)': 3.5,
 'Opera (1987)': 3.5,
 'Immortals (2011)': 3.39,
 'Skin I Live In, The (La piel que habito) (2011)': 3.38,
 'Host, The (2013)': 3.38,
 'Phantasm II (1988)': 3.17,
 'Holiday (1938)': 3.17,
 'Exorcist III, The (1990)': 3.12,
 'Clonus Horror, The (1979)': 3.06,
 'The Purge: Election Year (2016)': 3.06,
 "To Grandmother's House We Go (1992)": 3.06,
 'Despicable Me 3 (2017)': 3.06,
 'Garfield: A Tail of Two Kitties (2006)': 3.06,
 'I, Frankenstein (2014)': 3.06,
 'Room in Rome (Habitación en Roma) (2010)': 3.06,
 'Babylon 5': 3.06,
 'Young Adult (2011)': 3.06,
 'The Cloverfield Paradox (2018)': 3.06,
 'Wolfman, The (2010)':

Видно, что несколько известных фильмов исчезли из топ неоднозначных. Вывод: малые выборки следует использовать с осторожностью.

#### 7. Рейтинг фильма по movieId
Метод **movie_rating** возвращает рейтинг фильма по movieId, вычисленный по заданной метрике ('average', 'median', 'variance', 'count') из self.movie_ratings

Например, количество отзывов и средняя оценка Star Wars: Episode IV - A New Hope (1977):

In [33]:
%timeit -n 1 -r 1 ratings_full.movies.movie_rating(260, metric='count')
ratings_full.movies.movie_rating(260, metric='count')

5.64 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


251

In [34]:
%timeit -n 1 -r 1 ratings_full.movies.movie_rating(260, metric='average')
ratings_full.movies.movie_rating(260, metric='average')

75.7 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


4.23

## Часть 2.2. Класс Ratings.Users

Содержит статистику отзывов по пользователям. Наследуется от класса Ratings.Movies: все методы наследуются, методы с одинаковым названием переписываются.

#### 1. Топ пользователей по количеству оценок
Метод **top_by_num_of_ratings** вызывает метод self.top_by_ratings с параметром metric='count'.

In [35]:
%timeit -n 1 -r 1 ratings.users.top_by_num_of_ratings(10)
ratings.users.top_by_num_of_ratings(10)

172 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{6: 314, 1: 232, 4: 216, 7: 126, 5: 44, 3: 39, 2: 29}

Посмотрим статистику по всем 100k+ отзывам:

In [36]:
%timeit -n 1 -r 1 ratings_full.users.top_by_num_of_ratings(10)
ratings_full.users.top_by_num_of_ratings(10)

15.1 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{414: 2698,
 599: 2478,
 474: 2108,
 448: 1864,
 274: 1346,
 610: 1302,
 68: 1260,
 380: 1218,
 606: 1115,
 288: 1055}

Самые активные пользователи оставили по несколько тысяч отзывов.

#### 2. Топ пользователей по средней оценке
Метод **top_by_ratings** работает аналогично одноименному методу из Ratings.Movies, только вместо списка названий фильмов в self._top_x_by_y передается список self.userId.

In [37]:
%timeit -n 1 -r 1 ratings.users.top_by_ratings(10)
ratings.users.top_by_ratings(10)

133 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1: 4.37, 2: 3.95, 5: 3.64, 4: 3.56, 6: 3.49, 7: 3.35, 3: 2.44}

Повторим для более полной выборки:

In [38]:
%timeit -n 1 -r 1 ratings_full.users.top_by_ratings(10)
ratings_full.users.top_by_ratings(10)

16.4 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{53: 5.0,
 251: 4.87,
 515: 4.85,
 25: 4.81,
 30: 4.74,
 523: 4.69,
 348: 4.67,
 171: 4.63,
 452: 4.56,
 122: 4.55}

Всего один пользователь ставит только "пятерки". Вывод: пользователи оценивают не только понравившиеся фильмы.

#### 3. Топ пользователей по медианной оценке

In [39]:
%timeit -n 1 -r 1 ratings.users.top_by_ratings(10, 'median')
ratings.users.top_by_ratings(10, 'median')

210 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1: 5.0, 4: 4.0, 7: 4.0, 5: 4.0, 2: 4.0, 6: 3.0, 3: 0.5}

И для более полной выборки:

In [40]:
%timeit -n 1 -r 1 ratings_full.users.top_by_ratings(10, 'median')
ratings_full.users.top_by_ratings(10, 'median')

24.5 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1: 5.0,
 452: 5.0,
 52: 5.0,
 393: 5.0,
 43: 5.0,
 59: 5.0,
 171: 5.0,
 543: 5.0,
 523: 5.0,
 417: 5.0}

Пользователей, выставивших большинство "пятерок", ожидаемо много.

#### 4. Топ пользователей с неоднозначными оценками
Метод **top_controversial** вызывает метод self.top_by_ratings с параметром metric='variance'.

In [41]:
%timeit -n 1 -r 1 ratings.users.top_controversial(10)
ratings.users.top_controversial(10)

331 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{3: 4.26, 4: 1.72, 7: 1.65, 5: 0.96, 6: 0.72, 1: 0.64, 2: 0.63}

И, по традиции, для более полной выборки:

In [42]:
%timeit -n 1 -r 1 ratings_full.users.top_controversial(10)
ratings_full.users.top_controversial(10)

27.4 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{3: 4.26,
 461: 3.1,
 55: 3.09,
 259: 2.94,
 329: 2.92,
 502: 2.76,
 175: 2.75,
 598: 2.7,
 393: 2.61,
 138: 2.44}

Максимальные значения дисперсии оценок пользователей примерно такие же, как для фильмов. Но здесь это, скорее, говорит об объективности, чем о неоднозначности.

# Часть 3. Класс Tags

Создадим экземпляр класса:

In [43]:
%timeit -n 1 -r 1 tags = Tags('../datasets/tags.csv', 1000)
tags = Tags('../datasets/tags.csv', 1000)

10.7 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Атрибуты класса:

In [44]:
%timeit -n 1 -r 1 attr_list(tags)
attr_list(tags)

5.61 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['N',
 'COLS',
 'userId',
 'movieId',
 'tag',
 'date_time',
 'tag_count',
 'word_index']

N - количество фильмов, COLS - названия колонок данных. Списки userId, movieId, tag, date_time содержат данные из tags.csv: userId, movieId, теги, дата-время создания тегов, соответственно. tag_count - словарь уникальных тегов и их количество; word_index - словарь {слово: список индексов}, содержащий индексы тегов, где встречается слово.  
Метод считывает и преобразует данные из tags.csv аналогично методу Movies (см. Часть 1).
tag_count создается с помощью collections.Counter из списка self.tag
word_index создается с помощью метода self._get_words, который делит тег на слова методом str.split и обрезает знаки препинания с помощью регулярного выражения в методе self._shave_word.

#### 1. Теги с наибольшим количеством слов
Метод **most_words** создает список кортежей (tag, количество_слов), сортирует по количеству слов (вычисляется с помощью метода self._get_words) и преобразует в список.

In [45]:
%timeit -n 1 -r 1 tags.most_words(10)
tags.most_words(10)

2.62 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'Everything you want is here': 5,
 'Oscar (Best Music - Original Score)': 5,
 'based on a true story': 5,
 'heroine in tight suit': 4,
 'lord of the rings': 4,
 'Guardians of the Galaxy': 4,
 'jay and silent bob': 4,
 'based on a book': 4}

В основном теги содержат до 4 слов, но есть пара "выдающихся". Посмотрим, насколько много таких тегов в более полной выборке:

In [46]:
%timeit -n 1 -r 1 tags_full = Tags('../datasets/tags.csv', 10000)
tags_full = Tags('../datasets/tags.csv', 10000)

29.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Сколько в ней содержится тегов?

In [47]:
%timeit -n 1 -r 1 len(tags_full.userId)
len(tags_full.userId)

1.73 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


3683

А сколько уникальных?

In [48]:
%timeit -n 1 -r 1 len(tags_full.tag_count)
len(tags_full.tag_count)

1.45 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


1589

В каких тегах больше всего слов?

In [49]:
%timeit -n 1 -r 1 tags_full.most_words(10)
tags_full.most_words(10)

4.64 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'villain nonexistent or not needed for good story': 8,
 'It was melodramatic and kind of dumb': 7,
 '06 Oscar Nominated Best Movie - Animation': 6,
 'stop using useless characters for filler': 6,
 'Everything you want is here': 5,
 'Oscar (Best Music - Original Score)': 5,
 'based on a true story': 5,
 'based on a TV show': 5}

Большинство тегов состоят из небольшого количества слов.

#### 2. Самые длинные теги
Метод **longest** создает список кортежей (tag, длина_тега), сортирует по длине тега и преобразует в список.

In [50]:
%timeit -n 1 -r 1 tags.longest(10)
tags.longest(10)

89.6 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'audience intelligence underestimated',
 'Oscar (Best Music - Original Score)',
 'assassin-in-training (scene)',
 'Oscar (Best Cinematography)',
 'Everything you want is here',
 'political right versus left',
 'representation of children',
 'Guardians of the Galaxy']

И для более полной выборки:

In [51]:
%timeit -n 1 -r 1 tags_full.longest(10)
tags_full.longest(10)

315 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 Oscar Nominated Best Movie - Animation',
 'stop using useless characters for filler',
 'Academy award (Best Supporting Actress)',
 'Oscar (Best Effects - Visual Effects)',
 'audience intelligence underestimated',
 'It was melodramatic and kind of dumb']

Как и следовало ожидать, самые длинные теги - в основном те же, что и с наибльшим количеством слов.

#### 3. Самые длинные теги с наибольшим количеством слов

Преобразованное в список пересечение множеств, созданных из результатов двух предыдущих методов.

In [52]:
%timeit -n 1 -r 1 tags.most_words_and_longest(10)
tags.most_words_and_longest(10)

2.39 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['the catholic church is the most corrupt organization in history',
 'Oscar (Best Music - Original Score)',
 'Guardians of the Galaxy',
 'Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'Everything you want is here']

Для более полной выборки:

In [53]:
%timeit -n 1 -r 1 tags_full.most_words_and_longest(10)
tags_full.most_words_and_longest(10)

5.59 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['villain nonexistent or not needed for good story',
 'Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'It was melodramatic and kind of dumb',
 '06 Oscar Nominated Best Movie - Animation',
 'the catholic church is the most corrupt organization in history',
 'stop using useless characters for filler']

Как и следовало ожидать, списки самых длинных и самых многословных тегов наполовину совпадают.

#### 4. Самые популярные теги
Метод **most_popular** выводит Counter self.tag_count, отсортированный по убыванию количества упоминаний тега, с помощью метода most_common.

In [54]:
%timeit -n 1 -r 1 tags.most_popular(10)
tags.most_popular(10)

106 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'funny': 15,
 'sci-fi': 14,
 'twist ending': 12,
 'dark comedy': 12,
 'atmospheric': 10,
 'superhero': 10,
 'comedy': 10,
 'action': 10,
 'suspense': 10,
 'Leonardo DiCaprio': 9}

Более полная выборка и список подлиннее:

In [55]:
%timeit -n 1 -r 1 tags_full.most_popular(30)
tags_full.most_popular(30)

311 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'In Netflix queue': 131,
 'atmospheric': 36,
 'superhero': 24,
 'thought-provoking': 24,
 'funny': 23,
 'Disney': 23,
 'surreal': 23,
 'religion': 22,
 'sci-fi': 21,
 'dark comedy': 21,
 'quirky': 21,
 'psychology': 21,
 'suspense': 20,
 'twist ending': 19,
 'visually appealing': 19,
 'crime': 19,
 'politics': 18,
 'music': 16,
 'time travel': 16,
 'mental illness': 16,
 'dark': 15,
 'comedy': 15,
 'aliens': 15,
 'space': 14,
 'mindfuck': 14,
 'dreamlike': 14,
 'black comedy': 13,
 'emotional': 13,
 'heist': 13,
 'anime': 12}

Много интересных тегов, есть из чего выбрать на любой вкус.

#### 5. Теги, содержащие определенное слово
Метод **tags_with** выводит преобразованное в список и отсортированное по алфавиту множество тегов, найденных с помощью индексации слов в self.word_index (см. выше).

In [56]:
%timeit -n 1 -r 1 tags.tags_with('ending')
tags.tags_with('ending')

23.5 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['ending',
 'great ending',
 'happy ending',
 'powerful ending',
 'stupid ending',
 'Stupid ending',
 'surprise ending',
 'twist ending',
 'Twist Ending']

И для более полной выборки:

In [57]:
%timeit -n 1 -r 1 tags_full.tags_with('ending')
tags_full.tags_with('ending')

17.4 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['confusing ending',
 'ending',
 'great ending',
 'happy ending',
 'powerful ending',
 'stupid ending',
 'Stupid ending',
 'surprise ending',
 'twist ending',
 'Twist Ending']

#### 6. Поиск в тегах по нескольким словам

Метод **tags_search** принимает список слов и возвращает список тегов, содержащих все эти слова (case-insensitive). Находит пересечение множеств индексов, созданных для каждого слова из self.word_index, с помощью reduce и функции set.intersection.

In [58]:
%timeit -n 1 -r 1 tags.tags_search(['star', 'wars'])
tags.tags_search(['star', 'wars'])

23.3 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['Star Wars', 'star wars']

#### 7. Самые популярные слова, встречающиеся в тегах
Метод **popular_words** проходит циклом по self.word_index и находит слова с самым длинным списком индексов тегов, содержащих эти слова.

In [59]:
%timeit -n 1 -r 1 tags.popular_words(10)
tags.popular_words(10)

244 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'comedy': 33,
 'dark': 23,
 'ending': 21,
 'sci-fi': 18,
 'funny': 17,
 'twist': 16,
 'action': 15,
 'atmospheric': 12,
 'book': 12,
 'will': 11}

Думаю, вы уже поняли:

In [60]:
%timeit -n 1 -r 1 tags_full.popular_words(20)
tags_full.popular_words(20)

547 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'in': 145,
 'netflix': 134,
 'queue': 131,
 'comedy': 64,
 'dark': 47,
 'atmospheric': 41,
 'and': 33,
 'ending': 30,
 'funny': 28,
 'space': 28,
 'sci-fi': 27,
 'visually': 27,
 'bad': 27,
 'crime': 26,
 'superhero': 25,
 'twist': 24,
 'music': 24,
 'disney': 24,
 'surreal': 24,
 'movie': 24}

Можно сделать вывод, что пользователи любят Netflix и черный юмор.

# Часть 4. Класс Links

Создадим экземпляр класса:

In [61]:
%timeit -n 1 -r 1 links = Links('../datasets/links.csv', 1000)
links = Links('../datasets/links.csv', 1000)

48.5 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Атрибуты класса:

In [62]:
%timeit -n 1 -r 1 attr_list(links)
attr_list(links)

4.35 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['N',
 'COLS',
 'movieId',
 'imdbId',
 'tmdbId',
 'movies',
 'movie_index',
 'avail_fields',
 'imdb_raw_data']

N - количество фильмов, COLS - названия колонок данных. Списки movieId, imdbId, tmdbId содержат данные из links.csv: movieId, imdbId (ID в базе фильмов IMDB) и tmdbId (ID в базе фильмов TMDB).  
movies - экземпляр класса Movies с данными о названиях фильмов из movies.csv (должен находиться в одной папке с links.csv). movie_index - словарь {movieId: index} для быстрого доступа к imdbId и tmdbId по movieId. avail_fields - список доступных для поиска полей. imdb_raw_data - кэш данных с сайта imdb - словарь с ключом imdbId, чтобы избежать повторной загрузки сайта при одинаковых запросах.  
Метод считывает и преобразует данные из links.csv аналогично методу Movies (см. Часть 1).

#### 1. Получение данных с сайта IMDB

Метод **get_imdb** позволяет получить информацию о фильме по movieId и списку полей. Метод проходит циклом for по списку movieId, загружая данные для соответствующего imdbId с сайта imdb.com при помощи метода self._parse_imdb или из кэша self.imdb_raw_data. Для каждого из переданных в аргументах полей применяется свое преобразование: текстовые поля с возможными несколькими значениями преобразуются в список с помощью str.split, бюджет и сборы - в int, время - из часов и минут в минуты (int).

Метод **_parse_imdb** отвечает за загрузку страницы для данного imdbID с помощью библиотеки requests и парсинг нужных полей с помощью BeautifulSoup. Загружаются поля с CSS-классами 'ipc-metadata-list__item' (элемент списка), 'ipc-metadata-list-item__label' (название поля) и 'ipc-metadata-list-item__list-content-item' (значения).

Доступные поля для поиска:

In [63]:
%timeit -n 1 -r 1 links.avail_fields
links.avail_fields

622 ns ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


['Budget',
 'Director',
 'Directors',
 'Country of origin',
 'Countries of origin',
 'Gross worldwide',
 'Runtime',
 'Stars']

Бюджет фильма, режиссер(ы), страна(ы) происхождения, общемировые сборы, продолжительность в минутах, известные актеры, соответственно.

Чтобы запросить данные о фильме, нужно знать movieId. Но можно использовать автоматически созданный экземпляр класса Movies, чтобы найти movieId по названию фильма:

In [64]:
%timeit -n 1 -r 1 links.movies.title_search(['lethal', 'weapon'])
links.movies.title_search(['lethal', 'weapon'])

5.69 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{1918: 'Lethal Weapon 4 (1998)',
 2000: 'Lethal Weapon (1987)',
 2001: 'Lethal Weapon 2 (1989)',
 2002: 'Lethal Weapon 3 (1992)'}

Отлично, теперь мы знаем movieId.

In [65]:
%timeit -n 1 -r 1 links.get_imdb([1918, 2000, 2001, 2002], ['Budget', 'Director', 'Country of origin', 'Gross worldwide', 'Runtime', 'Stars'])
links.get_imdb([1918, 2000, 2001, 2002], ['Budget', 'Director', 'Country of origin', 'Gross worldwide', 'Runtime', 'Stars'])

133 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[]

И... ничего. Видимо, все эти movieId не вошли в первую тысячу. Рассмотрим более полную версию:

In [66]:
%timeit -n 1 -r 1 links_full = Links('../datasets/links.csv', 10000)
links_full = Links('../datasets/links.csv', 10000)

57.9 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


In [67]:
%timeit -n 1 -r 1 len(links_full.movieId)
len(links_full.movieId)

1.26 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


9742

Записей столько же, сколько в movies. Повторим поиск:

In [68]:
%timeit -n 1 -r 1 links_full.get_imdb([1918, 2000, 2001, 2002], ['Budget', 'Director', 'Country of origin', 'Gross worldwide', 'Runtime', 'Stars'])
links_full.get_imdb([1918, 2000, 2001, 2002], ['Budget', 'Director', 'Country of origin', 'Gross worldwide', 'Runtime', 'Stars'])

5.91 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[[2002,
  35000000,
  ['Richard Donner'],
  ['United States'],
  321731527,
  118,
  ['Mel Gibson', 'Danny Glover', 'Joe Pesci']],
 [2001,
  28000000,
  ['Richard Donner'],
  ['United States'],
  227853986,
  114,
  ['Mel Gibson', 'Danny Glover', 'Joe Pesci']],
 [2000,
  15000000,
  ['Richard Donner'],
  ['United States'],
  120207127,
  109,
  ['Mel Gibson', 'Danny Glover', 'Gary Busey']],
 [1918,
  140000000,
  ['Richard Donner'],
  ['United States'],
  285444603,
  127,
  ['Mel Gibson', 'Danny Glover', 'Joe Pesci']]]

Все 4 фильма нашлись. Загрузка IMDB страниц занимает время, примерно по 2 секунды на запрос.

#### 2. Топ режиссеров по количеству фильмов
Метод **top_directors** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит циклом for по списку загруженных данных, и с помощью collections.Counter подсчитывает количество упоминаний каждого режиссера.

In [69]:
%timeit -n 1 -r 1 links.top_directors(10)
links.top_directors(10)

Request error while parsing imdbId = 0118114: 404 Client Error: Not Found for url: https://www.imdb.com/title/tt0118114/
30min 21s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Alfred Hitchcock': 11,
 'Woody Allen': 8,
 'Frank Capra': 6,
 'Stanley Kubrick': 6,
 'Robert Stevenson': 6,
 'Steven Spielberg': 5,
 'Rob Reiner': 5,
 'James Cameron': 5,
 'Martin Scorsese': 5,
 'John Carpenter': 5}

Лидирует Альфред Хичкок, снявший 11 из первой тысячи фильмов коллекции.

#### 3. Топ самых дорогих фильмов
Метод **most_expensive** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит циклом for по списку загруженных данных и добавляет их в список кортежей (название_фильма, бюджет), сортирует по убыванию бюджета и преобразует в словарь.  
Данные о названиях фильмов содержатся в автоматически созданном экземпляре класса Movies.

In [70]:
%timeit -n 1 -r 1 links.most_expensive(10)
links.most_expensive(10)

13.6 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Waterworld (1995)': 175000000,
 'True Lies (1994)': 115000000,
 'Terminator 2: Judgment Day (1991)': 102000000,
 'Eraser (1996)': 100000000,
 'Hunchback of Notre Dame, The (1996)': 100000000,
 'Batman Forever (1995)': 100000000,
 'Cutthroat Island (1995)': 98000000,
 'Twister (1996)': 92000000,
 'Judge Dredd (1995)': 90000000,
 'Die Hard: With a Vengeance (1995)': 90000000}

Самый дорогой фильм из первой тысячи - "Водный мир" (бюджет 175 млн. долларов), один из самых дорогих фильмов на момент выхода (1995 год).  
Считается, что самый дорогой фильм в истории кино - четвертые "Пираты Карибского моря" (2011). Сделаем поиск по названию фильма:

In [71]:
%timeit -n 1 -r 1 links.movies.title_search(['pirates', '2011'])
links.movies.title_search(['pirates', '2011'])

7.6 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{86880: 'Pirates of the Caribbean: On Stranger Tides (2011)'}

Бюджет и общемировые сборы:

In [72]:
%timeit -n 1 -r 1 links_full.get_imdb([86880], ['Budget', 'Gross worldwide'])
links_full.get_imdb([86880], ['Budget', 'Gross worldwide'])

2.06 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


[[86880, 250000000, 1046721266]]

Бюджет почти в полтора раза выше, чем у самого дорогого фильма из первой тысячи, что говорит в пользу использования более полной выборки.  
Видим, что фильм довольно прибыльный, общемировые сборы примерно в 4 раза превышают бюджет. Интересно, какой у него пользовательский рейтинг?

In [73]:
%timeit -n 1 -r 1 ratings_full.movies.movie_rating(86880)
ratings_full.movies.movie_rating(86880)

32.9 µs ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


3.23

No comments.

#### 4. Топ самых прибыльных фильмов
Метод **most_profitable** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит циклом for по списку загруженных данных и добавляет их в список кортежей (название_фильма, сборы - бюджет), сортирует по убыванию прибыли и преобразует в словарь.  
Данные о названиях фильмов содержатся в автоматически созданном экземпляре класса Movies.

In [74]:
%timeit -n 1 -r 1 links.most_profitable(10)
links.most_profitable(10)

18.3 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Jurassic Park (1993)': 1040096466,
 'Lion King, The (1994)': 934161632,
 'E.T. the Extra-Terrestrial (1982)': 786807407,
 'Star Wars: Episode IV - A New Hope (1977)': 764398507,
 'Independence Day (a.k.a. ID4) (1996)': 742400891,
 'Forrest Gump (1994)': 623226465,
 'Star Wars: Episode V - The Empire Strikes Back (1980)': 532016086,
 'Ghost (1990)': 483703557,
 'Aladdin (1992)': 476050219,
 'Home Alone (1990)': 459383642}

Наибольшую прибыль (более миллиарда долларов) среди первой тысячи фильмов принес "Парк Юрского Периода".

#### 5. Самые продолжительные фильмы
Метод **longest** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит циклом for по списку загруженных данных и добавляет их в список кортежей (название_фильма, продолжительность [мин]), сортирует по убыванию продолжительности и преобразует в словарь.
Данные о названиях фильмов содержатся в автоматически созданном экземпляре класса Movies.

In [75]:
%timeit -n 1 -r 1 links.longest(10)
links.longest(10)

17.7 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Gone with the Wind (1939)': 238,
 'Once Upon a Time in America (1984)': 229,
 'Lawrence of Arabia (1962)': 227,
 'Ben-Hur (1959)': 212,
 'Godfather: Part II, The (1974)': 202,
 'Giant (1956)': 201,
 "Schindler's List (1993)": 195,
 'Right Stuff, The (1983)': 193,
 'Nixon (1995)': 192,
 'Gandhi (1982)': 191}

Самый продолжительный фильм (4 часа) - "Унесенные ветром" 1939 года. "Однажды в Америке" 1984 года отстает от него всего на 10 минут.

#### 6. Самая дорогая минута экранного времени
Метод **top_cost_per_minute** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит циклом for по списку загруженных данных и добавляет их в список кортежей (название_фильма, бюджет / продолжительность), сортирует по убыванию стоимости минуты и преобразует в словарь.  
Данные о названиях фильмов содержатся в автоматически созданном экземпляре класса Movies.

In [76]:
%timeit -n 1 -r 1 links.top_cost_per_minute(10)
links.top_cost_per_minute(10)

15.8 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Waterworld (1995)': 1296296.3,
 'Hunchback of Notre Dame, The (1996)': 1098901.1,
 'Judge Dredd (1995)': 937500.0,
 'Space Jam (1996)': 909090.91,
 'Eraser (1996)': 869565.22,
 'Batman Forever (1995)': 826446.28,
 'True Lies (1994)': 815602.84,
 'Twister (1996)': 814159.29,
 'Cutthroat Island (1995)': 790322.58,
 'Terminator 2: Judgment Day (1991)': 744525.55}

Как и по абсолютному значению бюджета, лидирует "Водный мир".

#### 7. Топ стран по количеству фильмов
Метод **top_countries** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит двойным циклом for по спискам стран в списке загруженных данных и добавляет их в collections.Counter, извлекая топ-n с помощью метода most_common.

In [77]:
%timeit -n 1 -r 1 links.top_countries(10)
links.top_countries(10)

9.66 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'United States': 839,
 'United Kingdom': 168,
 'France': 109,
 'Canada': 41,
 'Italy': 40,
 'Germany': 37,
 'Japan': 31,
 'Australia': 15,
 'Ireland': 14,
 'Hong Kong': 11}

Как и следовало ожидать, большинство фильмов снято в США. Из европейских стран лидирует Великобритания, из азиатских - Япония.

#### 8. Топ кинозвезд по количеству фильмов
Метод **top_stars** загружает список данных imdb страниц с помощью get_imdb для всего списка self.movieId, затем проходит двойным циклом for по спискам кинозвезд в списке загруженных данных и добавляет их в collections.Counter, извлекая топ-n с помощью метода most_common.

In [78]:
%timeit -n 1 -r 1 links.top_stars(10)
links.top_stars(10)

5.1 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


{'Robert De Niro': 13,
 'Cary Grant': 10,
 'Gene Hackman': 10,
 'Harrison Ford': 9,
 'Robin Williams': 8,
 'Robert Duvall': 8,
 'Harvey Keitel': 8,
 'Bruce Willis': 8,
 'Val Kilmer': 7,
 'Sean Connery': 7}

Среди кинозвезд чаще всего появляется в первой тысяче фильмов Роберт де Ниро.

### _Спасибо за внимание!_